File: symmetric-core/src/main/java/org/jumpmind/symmetric/transport/file/FileTransportManager.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.symmetric.transport.file;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.symmetric.ISymmetricEngine;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.model.IncomingBatch;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.transport.IIncomingTransport;
import org.jumpmind.symmetric.transport.IOutgoingWithResponseTransport;
import org.jumpmind.symmetric.transport.ITransportManager;
import org.jumpmind.symmetric.transport.http.HttpTransportManager;
import org.jumpmind.util.FormatUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FileTransportManager extends HttpTransportManager implements ITransportManager {

    static final Logger log = LoggerFactory.getLogger(FileTransportManager.class);

    IParameterService parameterService;
    
    public FileTransportManager(ISymmetricEngine engine) {
        super(engine);
        this.parameterService = engine.getParameterService();
    }

    @Override
    public int sendAcknowledgement(Node remote, List<IncomingBatch> list, Node local, String securityToken, String registrationUrl)
            throws IOException {
        return HttpURLConnection.HTTP_OK;
    }

    @Override
    public void writeAcknowledgement(OutputStream out, Node remote, List<IncomingBatch> list, Node local, String securityToken)
            throws IOException {
    }

    @Override
    public IIncomingTransport getFilePullTransport(Node remote, Node local, String securityToken, Map<String, String> requestProperties,
            String registrationUrl) throws IOException {
        return getPullTransport(remote, local, securityToken, requestProperties, registrationUrl);
    }

    @Override
    public IOutgoingWithResponseTransport getFilePushTransport(Node remote, Node local, String securityToken, String registrationUrl)
            throws IOException {
        return getPushTransport(remote, local, securityToken, registrationUrl);
    }

    @Override
    public IIncomingTransport getPullTransport(Node remote, Node local, String securityToken, Map<String, String> requestProperties,
            String registrationUrl) throws IOException {
        return new FileIncomingTransport(remote, local, 
                getDirName(ParameterConstants.NODE_OFFLINE_INCOMING_DIR, local), 
                getDirName(ParameterConstants.NODE_OFFLINE_ARCHIVE_DIR, local),
                getDirName(ParameterConstants.NODE_OFFLINE_ERROR_DIR, local));
    }

    @Override
    public IOutgoingWithResponseTransport getPushTransport(Node remote, Node local, String securityToken, String registrationUrl)
            throws IOException {   
        return new FileOutgoingTransport(remote, local, getDirName(ParameterConstants.NODE_OFFLINE_OUTGOING_DIR, local));
    }
    
    protected String getDirName(String paramName, Node localNode) {
        String dirName = parameterService.getString(paramName);
        dirName = FormatUtils.replace("nodeGroupId", localNode.getNodeGroupId(), dirName);
        dirName = FormatUtils.replace("nodeId", localNode.getNodeId(), dirName);        
        if (StringUtils.isNotBlank(dirName)) {
            File dir = new File(dirName);
            if (!dir.exists()) {
                dir.mkdirs();
            }
        }
        return dirName;
    }

}

File: symmetric-client-clib/src/transport/file/FileTransportManager.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "transport/file/FileTransportManager.h"

int SymFileTransportManager_sendAcknowledgement(SymFileTransportManager *this, SymNode *remote, SymList *batches, SymNode *local, char *securityToken, char *registrationUrl) {
    return SYM_TRANSPORT_OK;
}

char * SymFileTransportManager_getDirName(SymFileTransportManager *this, char *paramName, SymNode *localNode, char *defaultDirName) {
    SymTransportManager *super = &this->super;
    char *dirName = super->parameterService->getString(super->parameterService, paramName, defaultDirName);
    return dirName;
}

SymFileIncomingTransport * SymFileTransportManager_getPullTransport(SymFileTransportManager *this, SymNode *remote, SymNode *local, char *securityToken, SymProperties *requestProperties, char *registrationUrl) {
     return SymFileIncomingTransport_new(NULL, remote, local,
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_INCOMING_DIR, local, "./tmp/offline/incoming"),
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_ARCHIVE_DIR, local, ""),
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_ERROR_DIR, local, ""));
}

SymOfflineFileSyncIncomingTransport * SymFileTransportManager_getFileSyncPullTransport(SymFileTransportManager *this, SymNode *remote, SymNode *local, char *securityToken, SymProperties *requestProperties, char *registrationUrl) {
     return SymOfflineFileSyncIncomingTransport_new(NULL, remote, local,
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_INCOMING_DIR, local, "./tmp/offline/incoming"),
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_ARCHIVE_DIR, local, ""),
             SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_ERROR_DIR, local, ""));
}

SymFileOutgoingTransport * SymFileTransportManager_getPushTransport(SymFileTransportManager *this, SymNode *remote, SymNode *local, char *securityToken, char *registrationUrl) {
    return SymFileOutgoingTransport_new(NULL, remote, local,
            SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_OUTGOING_DIR, local, "./tmp/offline/outgoing"));
}

SymOfflineFileSyncOutgoingTransport * SymFileTransportManager_getFileSyncPushTransport(SymFileTransportManager *this, SymNode *remote, SymNode *local, char *securityToken, char *registrationUrl) {
    return SymOfflineFileSyncOutgoingTransport_new(NULL, remote, local,
            SymFileTransportManager_getDirName(this, SYM_PARAMETER_NODE_OFFLINE_OUTGOING_DIR, local, "./tmp/offline/outgoing"));
}

SymList * SymFileTransportManager_readAcknowledgement(SymFileTransportManager *this, char *parameterString1, char *parameterString2) {
    SymStringBuilder *sb = SymStringBuilder_newWithString(parameterString1);
    sb->append(sb, "&")->append(sb, parameterString2);

    SymMap *parameters = SymHttpTransportManager_getParametersFromQueryUrl(parameterString1);
    SymList *batchAcks = SymList_new(NULL);
    SymStringArray *keys = parameters->keys(parameters);
    int i;
    for (i = 0; i < keys->size; i++) {
        char *key = keys->get(keys, i);
        if (strncmp(key, SYM_WEB_CONSTANTS_ACK_BATCH_NAME, strlen(SYM_WEB_CONSTANTS_ACK_BATCH_NAME)) == 0) {
            char *batchId = SymStringUtils_substring(key, strlen(SYM_WEB_CONSTANTS_ACK_BATCH_NAME), strlen(key));
            SymBatchAck *batchAck = SymHttpTransportManager_getBatchInfo(parameters, batchId);
            batchAcks->add(batchAcks, batchAck);
            free(batchId);
        }
    }
    keys->destroy(keys);
    parameters->destroy(parameters);
    sb->destroy(sb);
    return batchAcks;
}

void SymFileTransportManager_destroy(SymFileTransportManager *this) {
    free(this);
}

SymFileTransportManager * SymFileTransportManager_new(SymFileTransportManager *this, SymParameterService *parameterService) {
    if (this == NULL) {
        this = (SymFileTransportManager *) calloc(1, sizeof(SymFileTransportManager));
    }
    SymTransportManager *super = &this->super;
    super->parameterService = parameterService;
    super->sendAcknowledgement = (void *) &SymFileTransportManager_sendAcknowledgement;
    super->getPullTransport = (void *) &SymFileTransportManager_getPullTransport;
    super->getPushTransport = (void *) &SymFileTransportManager_getPushTransport;
    super->getFileSyncPullTransport = (void *) &SymFileTransportManager_getFileSyncPullTransport;
    super->getFileSyncPushTransport = (void *) &SymFileTransportManager_getFileSyncPushTransport;
    super->readAcknowledgement = (void *) SymFileTransportManager_readAcknowledgement;
    super->destroy = (void *) &SymFileTransportManager_destroy;
    return this;
}
